[ORCA-3467] Orchestration and path resource documentation (#501)
PagerDuty/terraform-provider-pagerduty: website/docs/r/event_orchestration_unrouted.html.markdown

---
layout: "pagerduty"
page_title: "PagerDuty: pagerduty_event_orchestration_unrouted"
sidebar_current: "docs-pagerduty-resource-event-orchestration-unrouted"
description: |-
  Creates and manages an Unrouted Orchestration for a Global Event Orchestration in PagerDuty.
---

# pagerduty_event_orchestration_unrouted

An Unrouted Orchestration allows users to create a set of Event Rules that will be evaluated against all events that don't match any rules in the Orchestration's Router.

The Unrouted Orchestration evaluates events sent to it against each of its rules, beginning with the rules in the "start" set. When a matching rule is found, it can modify and enhance the event and can route the event to another set of rules within this Unrouted Orchestration for further processing.

## Example of configuring Unrouted Rules for an Orchestration

In this example of an Unrouted Orchestration, the rule matches only if the condition is matched.
Alerts created for events that do not match the rule will have severity level set to `info` as defined in `catch_all` block.

> `catch_all` for Unrouted Orchestration always suppressess the alert so it is not required to define the `suppress` action.

Reviewer: I'd argue it's more like "events that are sent to the Unrouted Orchestration will never belong to a service & therefore will **always** be suppressed, `catch_all` or not!" I'm going to remove this sentence for now, but feel free to add it (or a variation) back if you think it's removal makes things unclear. 👍 

```suggestion
```

---
layout: "pagerduty"
page_title: "PagerDuty: pagerduty_event_orchestration_unrouted"
sidebar_current: "docs-pagerduty-resource-event-orchestration-unrouted"
description: |-
  Creates and manages an Unrouted Orchestration for a Global Event Orchestration in PagerDuty.
---

# pagerduty_event_orchestration_unrouted

An Unrouted Orchestration allows users to create a set of Event Rules that will be evaluated against all events that don't match any rules in the Orchestration's Router.

The Unrouted Orchestration evaluates events sent to it against each of its rules, beginning with the rules in the "start" set. When a matching rule is found, it can modify and enhance the event and can route the event to another set of rules within this Unrouted Orchestration for further processing.

## Example of configuring a Unrouted Rules for an Orchestration

>   In this example the user has defined the unrouted orchestration with two rules, each routing to a different service.This example assumes services used in the route_to configuration already exists. So it does not show creation of service resource.As there is a single set, route_to is not defined for the rules.
> 

Reviewer: **Blocker:** "two rules, each routing to a different service" I don't think it's possible to route to a service from the Unrouted Orchestration. Honestly, I think this whole sentence should go away.

```suggestion
```

>   If there are more than one set, the rules in start set must define route_to with id of the next set
>   The first rule matches only if the condition is matched. The second rule matches always as there are no conditions.

Reviewer: **Not a Blocker:** Since this example doesn't demonstrate nested rules, I don't think we need to mention nested rules in this sentence specifically. Maybe something like this instead?

```suggestion
  In this example of an Unrouted Orchestration, first rule matches only if the condition is matched. The second rule matches all remaining events because it doesn't have any conditions.
```

>   `catch_all` with empty `actions` block suppresses the alerts that do not match any rules. Using the `catch_all` to set severity, event_action, variables and extractions is also allowed.

Reviewer: **Blocker:** Again, I don't think suppress-by-default is correct (see my comment above).

Also, looking over this example (which is the same as [the API Docs example](https://developer.pagerduty.com/api-reference/953ddc1761403-update-the-unrouted-orchestration-for-a-global-event-orchestration)) makes me realize we have a bug! Seems like we should just use the `catch_all` block _instead_ of creating a rule with no conditions! In fact, that 2nd condition-less rule means the catch_all will **never** be used!

Might be worth changing this example to actually use the `catch_all` instead of 2 rules just to show how it can be used?

Author: Good call out. It is true that for service path, the catch all behaviour leaves event as-is. But for `unrouted` the `catch all `always suppresses alert.
<img width="1422" alt="image" src="https://user-images.githubusercontent.com/3372372/169693979-71314b28-74ef-47d7-8572-0c5b69889416.png">
But I do agree that the example is confusing in the sense that 2nd rule will always match so catch_all will never be hit. I have removed the 2nd rule and moved it's alert action to the `catch_all`

```hcl
resource "pagerduty_event_orchestration_unrouted" "unrouted" {
 event_orchestration = pagerduty_event_orchestration.my_monitor.id
  sets {
    id = "start"
    rules {
      label = "Update the summary of un-matched Critical alerts so they're easier to spot"
      conditions {
        expression = "event.severity matches 'critical'"
      }
      actions {
        severity = "info"
        extractions {

>             target = "event.summary"
>             template = "[Critical Unrouted] {{event.summary}}"

Reviewer: **Nit:** too deeply indented

```suggestion
          target = "event.summary"
          template = "[Critical Unrouted] {{event.summary}}"
```

        }
      }
    }
    rules {
      label = "Reduce the severity of all other unrouted events"
      actions {
        severity = "info"
      }
    }
  }
  catch_all {
    actions { }
  }
}
```

## Argument Reference

The following arguments are supported:

* `event_orchestration` - (Required) The Event Orchestration to which this Unrouted Orchestration belongs to.
* `sets` - (Required) An Unrouted Orchestration must contain at least a "start" set, but can contain any number of additional sets that are routed to by other rules to form a directional graph.

> * `catch_all` - (Required) When none of the rules match an event, the event will be routed according to the catch_all settings.

Reviewer: **Blocker:** Same as above, in an Unrouted Orchestration the `catch_all` cannot do routing. Maybe this instead?

```suggestion
* `catch_all` - (Required) the `catch_all` actions will be applied if an Event reaches the end of any set without matching any rules in that set.
```

### Sets (`sets`) supports the following:
* `id` - (Required) The ID of this set of rules. Rules in other sets can route events into this set using the rule's `route_to` property.

> * `rules` - (Optional) The Unrouted Orchestration evaluates Events against these Rules, one at a time, and routes each Event based on the first rule that matches. If no rules are provided as part of Terraform configuration, the API returns empty list of rules.

Reviewer: **Blocker:** This isn't a router. Maybe this instead?

```suggestion
* `rules` - (Optional) The Unrouted Orchestration evaluates Events against these Rules, one at a time, and applies all the actions for first rule it finds where the event matches the rule's conditions. If no rules are provided as part of Terraform configuration, the API returns empty list of rules.
```

### Rules (`rules`) supports the following:
* `label` - (Optional) A description of this rule's purpose.
* `conditions` - (Optional) Each of these conditions is evaluated to check if an event matches this rule. The rule is considered a match if any of these conditions match. If none are provided, the event will `always` match against the rule.
* `actions` - (Required) Actions that will be taken to change the resulting alert and incident, when an event matches this rule.
* `disabled` - (Optional) Indicates whether the rule is disabled and would therefore not be evaluated.

### Conditions (`conditions`) supports the following:
* `expression`- (Required) A [PCL condition](https://developer.pagerduty.com/docs/ZG9jOjM1NTE0MDc0-pcl-overview) string.

### Actions (`actions`) supports the following:

> * `route_to` - (Required) The ID of the target Service for the resulting alert.

Reviewer: **Blocker:** Nope, that's the router wording. Also it's Optional. Maybe this instead, taken from the REST API docs?

```suggestion
* `route_to` - (Optional) The ID of a Set from this Unrouted Orchestration whose rules you also want to use with event that match this rule.
```
